### PR TITLE
Added Supervisor/Server to Scanner

### PR DESCRIPTION
**Pull request by nomicflux**
Added new `ScannerSupervisor` and `ScannerServer` modules.  `ScannerSupervisor` starts `ScannerServer`, `ScannerServer` then has `ScannerSupervisor` start `Scanner`, and monitors for `Scanner` for exit.  Currently it just logs the exit reason, but could do whatever is needed to notify and recover the `Scanner`.

### CODE REVIEW

File: lib/mke_police.ex [hunk 1]
```diff
@@ -14,7 +14,8 @@ defmodule MkePolice do
       supervisor(MkePolice.Endpoint, []),
       # Start your own worker by calling: MkePolice.Worker.start_link(arg1, arg2, arg3)
       # worker(MkePolice.Worker, [arg1, arg2, arg3]),
-      supervisor(MkePolice.Scanner, [5000])
+      supervisor(MkePolice.ScannerSupervisor, [5000])
+      # supervisor(MkePolice.Scanner, [5000])
```
Comment (jeregrine) [inline lib/mke_police.ex]: Remove commented code.

File: mix.lock [hunk 2]
```diff
@@ -1,36 +1,36 @@
...
-  "postgrex": {:hex, :postgrex, "0.12.1", "2f8b46cb3a44dcd42f42938abedbfffe7e103ba4ce810ccbeee8dcf27ca0fb06", [:mix], [{:connection, "~> 1.0", [hex: :connection, optional: false]}, {:db_connection, "~> 1.0-rc.4", [hex: :db_connection, optional: false]}, {:decimal, "~> 1.0", [hex: :decimal, optional: false]}]},
-  "ranch": {:hex, :ranch, "1.2.1", "a6fb992c10f2187b46ffd17ce398ddf8a54f691b81768f9ef5f461ea7e28c762", [:make], []},
-  "ssl_verify_fun": {:hex, :ssl_verify_fun, "1.1.0", "edee20847c42e379bf91261db474ffbe373f8acb56e9079acb6038d4e0bf414f", [:rebar, :make], []},
-  "timex": {:hex, :timex, "3.1.0", "71c1bc6eef88f2bf8628224d62eab12dda5c5904e9db764289db1e360c622900", [:mix], [{:combine, "~> 0.7", [hex: :combine, optional: false]}, {:gettext, "~> 0.10", [hex: :gettext, optional: false]}, {:tzdata, "~> 0.1.8 or ~> 0.5", [hex: :tzdata, optional: false]}]},
-  "tzdata": {:hex, :tzdata, "0.5.9", "575be217b039057a47e133b72838cbe104fb5329b19906ea4e66857001c37edb", [:mix], [{:hackney, "~> 1.0", [hex: :hackney, optional: false]}]}}
+%{"certifi": {:hex, :certifi, "0.7.0", "861a57f3808f7eb0c2d1802afeaae0fa5de813b0df0979153cbafcd853ababaf", [:rebar3], []},
```
Comment (jeregrine) [inline mix.lock]: this file changed bug mix.exs did not.

File: lib/mke_police/scanner_server.ex [hunk 3]
```diff
@@ -0,0 +1,56 @@
+defmodule MkePolice.ScannerServer do
+  require Logger
+  use GenServer
+  import Supervisor.Spec
+
+  @name :"MkePolice.ScannerServer"
```
Comment (jeregrine) [inline lib/mke_police/scanner_server.ex]: would give his a better name, possibly a symbol or `__MODULE__`

File: lib/mke_police/scanner.ex [hunk 4]
```diff
@@ -1,13 +1,17 @@
-defmodule MkePolice.Scanner do 
+defmodule MkePolice.Scanner do
 
   alias MkePolice.{Repo, Call}
-  
 
-  def start_link(restart_interval) do 
-    GenServer.start_link(__MODULE__, restart_interval)
+  @name :"MkePolice.Scanner"
```
Comment (jeregrine) [inline lib/mke_police/scanner.ex]: would give his a better name, possibly a symbol or __MODULE__

File: mix.lock (hunk 2 above)
Comment (nomicflux) [inline mix.lock]: :certifi added to mix.exs

**Comment by nomicflux** [comment]
After being restarted, Scanner terminates with the following:
```
[error] Scanner died: {:undef, [{MkePolice.Scanner, :terminate, [{%Ecto.ConstraintError{constraint: "calls_pkey", message: "constraint error when attempting to insert struct:\n\n    * unique: calls_pkey\n\nIf you would like to convert this constraint into an error, please\ncall unique_constraint/3 in your changeset and define the proper\nconstraint name. The changeset has not defined any constraint.\n", type: :unique}, [{Ecto.Repo.Schema, :"-constraints_to_errors/3-fun-1-", 4, [file: 'lib/ecto/repo/schema.ex', line: 423]}, {Enum, :"-map/2-lists^map/1-0-", 2, [file: 'lib/enum.ex', line: 1184]}, {Ecto.Repo.Schema, :constraints_to_errors, 3, [file: 'lib/ecto/repo/schema.ex', line: 409]}, {Ecto.Repo.Schema, :"-do_insert/4-fun-0-", 11, [file: 'lib/ecto/repo/schema.ex', line: 201]}, {Ecto.Repo.Schema, :insert!, 4, [file: 'lib/ecto/repo/schema.ex', line: 124]}, {MkePolice.Scanner, :"-handle_info/2-fun-1-", 1, [file: 'lib/mke_police/scanner.ex', line: 37]}, {Enum, :"-each/2-lists^foreach/1-0-", 2, [file: 'lib/enum.ex', line: 651]}, {Enum, :each, 2, [file: 'lib/enum.ex', line: 651]}, {MkePolice.Scanner, :handle_info, 2, [file: 'lib/mke_police/scanner.ex', line: 24]}, {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]}, {:gen_server, :handle_msg, 5, [file: 'gen_server.erl', line: 667]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}]}, %MkePolice.Scanner.State{interval: 5000, server_pid: #PID<0.295.0>}], []}, {:gen_server, :try_terminate, 3, [file: 'gen_server.erl', line: 629]}, {:gen_server, :terminate, 7, [file: 'gen_server.erl', line: 795]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}]} (2016-11-22 01:00:17.641152Z)
[error] GenServer MkePolice.Scanner terminating
** (UndefinedFunctionError) function MkePolice.Scanner.terminate/2 is undefined or private
    (mke_police) MkePolice.Scanner.terminate({%Ecto.ConstraintError{constraint: "calls_pkey", message: "constraint error when attempting to insert struct:\n\n    * unique: calls_pkey\n\nIf you would like to convert this constraint into an error, please\ncall unique_constraint/3 in your changeset and define the proper\nconstraint name. The changeset has not defined any constraint.\n", type: :unique}, [{Ecto.Repo.Schema, :"-constraints_to_errors/3-fun-1-", 4, [file: 'lib/ecto/repo/schema.ex', line: 423]}, {Enum, :"-map/2-lists^map/1-0-", 2, [file: 'lib/enum.ex', line: 1184]}, {Ecto.Repo.Schema, :constraints_to_errors, 3, [file: 'lib/ecto/repo/schema.ex', line: 409]}, {Ecto.Repo.Schema, :"-do_insert/4-fun-0-", 11, [file: 'lib/ecto/repo/schema.ex', line: 201]}, {Ecto.Repo.Schema, :insert!, 4, [file: 'lib/ecto/repo/schema.ex', line: 124]}, {MkePolice.Scanner, :"-handle_info/2-fun-1-", 1, [file: 'lib/mke_police/scanner.ex', line: 37]}, {Enum, :"-each/2-lists^foreach/1-0-", 2, [file: 'lib/enum.ex', line: 651]}, {Enum, :each, 2, [file: 'lib/enum.ex', line: 651]}, {MkePolice.Scanner, :handle_info, 2, [file: 'lib/mke_police/scanner.ex', line: 24]}, {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]}, {:gen_server, :handle_msg, 5, [file: 'gen_server.erl', line: 667]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}]}, %MkePolice.Scanner.State{interval: 5000, server_pid: #PID<0.295.0>})
    (stdlib) gen_server.erl:629: :gen_server.try_terminate/3
    (stdlib) gen_server.erl:795: :gen_server.terminate/7
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: :scan
State: %MkePolice.Scanner.State{interval: 5000, server_pid: #PID<0.295.0>}
```

**Comment by jeregrine** [comment]
@nomicflux the error is hidden,
```
%Ecto.ConstraintError{constraint: "calls_pkey", message: "constraint error when attempting to insert struct:\n\n    * unique: calls_pkey\n\nIf you would like to convert this constraint into an error, please\ncall unique_constraint/3 in your changeset and define the proper\nconstraint name. The changeset has not defined any constraint.\n", type: :unique}, [{Ecto.Repo.Schema, :"-constraints_to_errors/3-fun-1-", 4, [file: 'lib/ecto/repo/schema.ex', line: 423]}, {Enum, :"-map/2-lists^map/1-0-", 2, [file: 'lib/enum.ex', line: 1184]}, {Ecto.Repo.Schema, :constraints_to_errors, 3, [file: 'lib/ecto/repo/schema.ex', line: 409]}, {Ecto.Repo.Schema, :"-do_insert/4-fun-0-", 11, [file: 'lib/ecto/repo/schema.ex', line: 201]}, {Ecto.Repo.Schema, :insert!, 4, [file: 'lib/ecto/repo/schema.ex', line: 124]}, {MkePolice.Scanner, :"-handle_info/2-fun-1-", 1, [file: 'lib/mke_police/scanner.ex', line: 37]}, {Enum, :"-each/2-lists^foreach/1-0-", 2, [file: 'lib/enum.ex', line: 651]}, {Enum, :each, 2, [file: 'lib/enum.ex', line: 651]}, {MkePolice.Scanner, :handle_info, 2, [file: 'lib/mke_police/scanner.ex', line: 24]}, {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]}, {:gen_server, :handle_msg, 5, [file: 'gen_server.erl', line: 667]}, {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}]}, %MkePolice.Scanner.State{interval: 5000, server_pid: #PID<0.295.0>})
```

**Comment by nomicflux** [comment]
I think that pull request #4 is reworking the Call changeset, so I'll hold off fixing this until that request is approved.

**Comment by nomicflux** [comment]
With adding terminate/2 to MkePolice.Scanner, the error messages went away.  MkePolice.Scanner can be killed in :observer.start; they die and are immediately restarted, with both events logged.

**Comment by nickgartmann** [comment]
This seems to work well, and looks good. Planning on merging (you have everything in that you want @nomicflux?)

**Comment by nomicflux** [comment]
@nickgartmann: Yes, it's a good starting point.  I want to get in tests next, but that should be its own PR.

**Comment by nickgartmann** [comment]
Thanks <3

**Pull request merged**